### PR TITLE
Improvement 2.x identitydetection 2023-01-24

### PR DESCRIPTION
**Pull request by dodmi**
I'm using local folders to archive messages; sent mails were marked with DKIM errors. So this change tries to identify all mails, which were sent from known identities (but not in junk)

### CODE REVIEW

**Review by lieser — COMMENTED** [review]
Thanks for the improved detection but the redirect feature is still a little unclear to me.

File: modules/AuthVerifier.jsm [hunk 1]
```diff
@@ -732,17 +734,62 @@ function isOutgoing(msgHdr) {
...
-	// return true if one of the servers identities contain the from address
-	let author = msgHdr.mime2DecodedAuthor;
-	let	identities = MailServices.accounts.
-			getIdentitiesForServer(msgHdr.folder.server);
-	for (let identity of fixIterator(identities, Ci.nsIMsgIdentity)) {
-		if (author.includes(identity.email)) {
-			return true;
+	// return true
+	// - if one of the servers identities contains the from address
+	// - if another account's inbox is redirected to this server and an originating account's identity matches
```
[the remark this answers is not in the document]
Comment (dodmi) [reply modules/AuthVerifier.jsm:744]: At least in TB52 you can redirect the inbox of a POP mail account to another one.
In this case, I'm adding the account's email addresses to the target account's addresses.

image

https://searchfox.org/comm-central/source/mailnews/base/public/nsIMsgIncomingServer.idl
/* the root folder for this server, even if server is deferred */
  attribute nsIMsgFolder rootFolder;
/* root folder for this account
     - if account is deferred, root folder of deferred-to account */
  readonly attribute nsIMsgFolder rootMsgFolder;

File: modules/AuthVerifier.jsm [hunk 2]
```diff
@@ -732,17 +734,62 @@ function isOutgoing(msgHdr) {
...
+	// - if the mail is in local folders and the from address matches any identity
+
+	const accMgr = MailServices.accounts;
+	const msgHeaderParser = Cc["@mozilla.org/messenger/headerparser;1"].createInstance(Ci.nsIMsgHeaderParser);
+	const lfAcc = accMgr.FindAccountForServer(accMgr.localFoldersServer);
+	const lfKey = lfAcc ? lfAcc.key : null;
+
+	let accountAddressDB = new Object;
```
[the remark this answers is not in the document]
Comment (dodmi) [reply modules/AuthVerifier.jsm]: Thank you for this improvement :)

File: modules/AuthVerifier.jsm (hunk 1 above)
[the remark this answers is not in the document]
Comment (lieser) [reply modules/AuthVerifier.jsm:744]: Thanks for the info, this seems to be a feature unique to POP Mail Server.

**Review by lieser — APPROVED** [review]
Thanks again for this, especially interesting to learn about the INBOX redirection for POP server. This could also be relevant then I will implement something similar for the current version (#114).

Note that the current algorithm could properly improved a little to avoid unnecessary work, e.g. by only collecting the identities for relevant accounts. I.e. if the INBOX is redirected or it is for the local folder.
But that would potentially complicated it more. So would be fine with merging as is, as I assume it is still fast enough to not really produce any noticeable delay.

**Comment by dodmi** [comment]
My productive installation has 8 different accounts and 20 identities configured. I don't notice any impact. 
For the local folders logic, I need to go through all accounts and collect the identities there, in any case. And this is the main cause, I've extended your implementation.
If there are users of the 2.x version experiencing problems, I'd gladly improve the algorithm. But for now, I'm fine with it.

**Pull request merged**